src/runtime/substitution.cc

```diff
@@ -1929,6 +1942,107 @@ void GraphSearchHelper::graph_optimize(
   optimal_views = real_optimal_views;
 }
 
+/**
+ * @brief Experimental DP algorithm to optimize PCG with the consideration of
+ * memory usage. This is to avoid polluting the current Unity search algorithm
+ * above. And this should be merged to GraphSearchHelper::graph_optimize
+ * eventually.
+ *
+ * @param[in] budget Not used
+ * @param[in] only_data_parallel Not used
+ * @param[out] best_graph The best possible PCG after optimization
+ * @param[out] optimal_views The corresponding device placement views of the
+ * best graph
+ * @param[out] search_result The performance result of the search
+ */
+void GraphSearchHelper::graph_optimize_with_memory(
+    size_t budget,
+    bool only_data_parallel,
+    std::unique_ptr<Graph> &best_graph,
+    std::unordered_map<Node, MachineView> &optimal_views,
+    MemorySearchResult &search_result) {
+  this->logger->debug()
+      << "Starting graph optimization with memory consideration";
+
+  // Construct graph structure
+  Graph *graph = this->construct_graph();
+
+  // The input nodes may need to be duplicated because the PCG was constructed
+  // to have one input node for one input, but the actual execution graph should
+  // have the distributed version of inputs (i.e. multiple nodes).
+  graph->duplicate_input_nodes();
+
+  // Export an empty schedule if needed.
+  std::unordered_map<Node, MachineView> empty_strategy;
+  if (!this->config.export_strategy_computation_graph_file.empty()) {
+    graph->export_strategy_computation_graph(
+        empty_strategy, this->config.export_strategy_computation_graph_file);
+  }
+
+  Node sink_node = graph->find_sink_node();
+
+  auto const start = std::chrono::system_clock::now();
+  GraphOptimizeResultWithMemory optimal =
+      this->generic_sequence_optimize_with_memory<
+          GraphOptimizeResultWithMemory>(
+          graph, sink_node, tl::nullopt, tl::nullopt);
+  auto const end = std::chrono::system_clock::now();
+
+  this->logger->debug() << "Total cache size: "
+                        << this->cached_optimized_graphs.size();
+  std::cout << "Optimal run time cost: " << optimal.cost
+            << ", Memory usage: " << optimal.mem_cost
+            << " | run_time_cost_factor: "
+            << this->mem_config.run_time_cost_factor << std::endl;
+
+  // Save the search performance results to the output argument
+  search_result.run_time_cost = optimal.cost;
+  search_result.memory_cost = optimal.mem_cost.num;
+  search_result.search_time =
+      std::chrono::duration_cast<std::chrono::milliseconds>(end - start)
+          .count();
+
+  // Further simplify the "optimal" graph/schedule to have a more efficient
+  // graph and more accurate cost.
+  best_graph = std::unique_ptr<Graph>(new Graph(optimal.graph.value()));
+  SimplificationSettings settings;
+  // Temp disable parallel_op fusion to observe the strategy in dot graph.
+  // TODO(eric): reverse this back
```

Comment:
?

Reply:
If we're merging TODOs into master they should be converted to issues and have an explanation of what the todo is and why it's there

Reply:
Updated.